wandb/integration/openai/fine_tune.py

```diff
@@ -0,0 +1,338 @@
+import datetime
+import io
+import os
+import json
+import time
+import re
+from pathlib import Path
+
+try:
+    import openai
+    from openai import OpenAI
+except ImportError as e:
+    raise Exception(
+        "Error: `openai` not installed >> This integration requires openai!  To fix, please `pip install --pre openai`"
+    ) from e
+
+try:
+    import numpy as np
+except ImportError as e:
+    raise Exception(
+        "Error: `numpy` not installed >> This integration requires numpy!  To fix, please `pip install numpy`"
+    ) from e
+
+try:
+    import pandas as pd
+except:
+    raise Exception(
+        "Error: `pandas` not installed >> This integration requires pandas!  To fix, please `pip install pandas`"
+    ) from e
+
+import wandb
+from wandb.sdk.lib import telemetry
+
+# from openai import File, FineTune, FineTuningJob
+
+
+class WandbLogger:
+    """
+    Log fine-tunes to [Weights & Biases](https://wandb.me/openai-docs)
+    """
+
+    _wandb_api = None
+    _logged_in = False
+    openai_client = None
+
+    @classmethod
+    def sync(
+        cls,
+        id=None,
+        n_fine_tunes=None,
+        project="OpenAI-Fine-Tune",
+        entity=None,
+        force=False,
+        blocking=True,
+        **kwargs_wandb_init,
+    ):
+        """
+        Sync fine-tunes to Weights & Biases.
+        :param id: The id of the fine-tune (optional)
+        :param n_fine_tunes: Number of most recent fine-tunes to log when an id is not provided. By default, every fine-tune is synced.
+        :param project: Name of the project where you're sending runs. By default, it is "GPT-3".
+        :param entity: Username or team name where you're sending runs. By default, your default entity is used, which is usually your username.
+        :param force: Forces logging and overwrite existing wandb run of the same fine-tune.
+        :param blocking: Waits for the fine-tune to be complete and then log metrics to W&B. By default, it is True.
+        """
+
+        openai_client = OpenAI(
+            api_key=os.environ['OPENAI_API_KEY'],
+        )
+        cls.openai_client = openai_client
+
+        if id:
+            print("Retrieving fine-tune job...")
+            fine_tune = openai_client.fine_tuning.jobs.retrieve(fine_tuning_job_id=id)
+        else:
+            # get list of fine_tune to log
+            fine_tunes = openai_client.fine_tuning.jobs.list()
+            if not fine_tunes or fine_tunes.data is None:
+                print("No fine-tune has been retrieved")
+                return
+            fine_tunes = fine_tunes.data[
+                -n_fine_tunes if n_fine_tunes is not None else None :
+            ]
+
+        # log starting from oldest fine_tune
+        show_individual_warnings = (
+            False if id is None and n_fine_tunes is None else True
+        )
+        fine_tune_logged = []
+        for fine_tune in fine_tunes:
```

Comment:
this won't work if `id` isn't passed right? in that case no `fine_tunes` is created

Reply:
If no `id` is passed, the else statement (`fine_tunes = openai_client.fine_tuning.jobs.list()`) will select all the jobs (completed and otherwise). It will then log the metrics for jobs which are not yet logged.

Reply:
you were right. added `fine_tunes = [fine_tune]` if `id` is available.